caffe2/opt/backend_cutting.cc

```diff
@@ -0,0 +1,416 @@
+// very first group, which can be all tensors
+struct TransformSubgraph {
+  explicit TransformSubgraph(
+      std::vector<NodeRef>&& f,
+      std::vector<NodeRef>&& n,
+      int id,
+      bool need)
+      : input_nodes(std::move(f)),
+        nodes(std::move(n)),
+        group_id(id),
+        needed(need) {}
+
+  TransformSubgraph(TransformSubgraph&& rhs) noexcept
+      : input_nodes(std::move(rhs.input_nodes)),
+        nodes(std::move(rhs.nodes)),
+        external_input_refs(std::move(rhs.external_input_refs)),
+        external_output_refs(std::move(rhs.external_output_refs)),
+        group_id(rhs.group_id),
+        needed(rhs.needed) {}
+
+  TransformSubgraph& operator=(TransformSubgraph&& rhs) noexcept {
+    input_nodes = std::move(rhs.input_nodes);
+    nodes = std::move(rhs.nodes);
+    external_input_refs = std::move(external_input_refs);
+    external_output_refs = std::move(external_output_refs);
+    group_id = rhs.group_id;
+    needed = rhs.needed;
+    return *this;
+  }
+
+  void Print() const {
+    LOG(INFO) << "Group :" << group_id;
+    LOG(INFO) << "  Input Nodes: ";
+    for (const auto i : input_nodes) {
+      LOG(INFO) << "    " << ShowNode(i);
+    }
+    LOG(INFO) << "  Nodes: ";
+    for (const auto i : nodes) {
+      LOG(INFO) << "    " << ShowNode(i);
+    }
+  }
+
+  std::vector<NodeRef> input_nodes;
+  std::vector<NodeRef> nodes;
+  std::unordered_map<std::string, NodeRef> external_input_refs;
+  std::unordered_map<std::string, NodeRef> external_output_refs;
+  int group_id{-1};
+  bool needed{true};
+};
+
+// Merge subgraph
+// Precondition: subgraphs appear in topological order in the list
+void MergeSubgraph(
+    std::vector<TransformSubgraph>* subs,
+    std::unordered_map<NodeRef, GroupAnnotation>* infos) {
+
+
+}
+
+caffe2::NetDef ConvertToC2Net(
+    const TransformSubgraph& sub,
+    const std::unordered_map<NodeRef, GroupAnnotation>& infos) {
+  caffe2::NetDef net;
+  for (auto node : sub.nodes) {
+    auto* node_data = node->data().get();
+    if (isa<nom::repr::NeuralNetOperator>(node_data)) {
+      const auto* nn_op =
+          dyn_cast<const nom::repr::NeuralNetOperator>(node_data);
+      const auto* op_def = reinterpret_cast<const caffe2::OperatorDef*>(
+          nn_op->getAnnotation()->getSaved());
+      net.add_op()->CopyFrom(*op_def);
+    } 
+  }
+  for (const auto kv : sub.external_input_refs) {
+    net.add_external_input(kv.first);
+    LOG(INFO) << "Adding external input: " << kv.first;
+  }
+  for (const auto& kv : sub.external_output_refs) {
+    net.add_external_output(kv.first);
+    LOG(INFO) << "Adding external output: " << kv.first;
+  }
+
+  return net;
+}
+
+void DetectBoundaryReferences(
+    TransformSubgraph* subgraph,
+    const std::unordered_map<NodeRef, GroupAnnotation>& infos) {
+  for (auto node: subgraph->nodes) {
+    auto* node_data = node->data().get();
+    // inputs
+    for (auto in_edge : node->getInEdges()) {
```

Comment:
perhaps this is better phrased as nodes containing `NeuralNetData` such that
`repr::nn::hasProducer(node) == false`

Reply:
Hmm, I do not get this. Could you clarify a bit?

Reply:
ah, my mistake -- I thought you were acting on `Graph`, not a `TransformSubgraph`